### ISSUE TITLE
Support for Bytes (Uint8Array)

### ISSUE THREAD

**Issue by nicprov** [billstclair/elm-s3#3]
Essentially, I'm trying to deal with encrypted objects. Dealing with strings seems to be causing issues as it later incorrectly gets converted to Uint8Array.

How would I go about receiving bytes directly without it first being converted to a string?

**Comment by billstclair** [comment]
I won't have time to look at this until tomorrow or Monday. It likely makes sense to allow `Bytes` for both `get` and `put` (via two new functions),  but I won't know for sure until I spend some time on it.

**Comment by nicprov** [comment]
Thanks for the reply and for taking the time to look into it! I do agree that adding two new methods would be ideal to ensure backwards compatibility. 

That being said, this change doesn't seem to be as straightforward as I though. The limitation seems to be more on the _elm-aws-core_ side. For some reason, every request seems to work with a _ResponseDecoder_ that is hard-coded to only work with strings as shown below:
`type alias ResponseDecoder a =
    Metadata -> String -> Result String a
`

The issue with _elm-aws-core_  is that it no longer seems to be maintained. There are four pull requests that have been there for over a year, so a fork may be necessary at this point.

Anyways, let me know if you can think of a clever way to do this!

**Comment by nicprov** [comment]
Ok, so I got it working (not in a remotely clean way though). I did kind of what we said and simply added a new method for bytes wherever necessary. 

For now, I added it to my repo, you can try downloading it here:
[elm-aws-core](https://github.com/nicprov/cryptzilla/tree/master/lib/elm-aws-core)
[elm-s3](https://github.com/nicprov/cryptzilla/tree/master/lib/elm-s3)

**Comment by nicprov** [comment]
I found one issue with uploading (putBytesObject) though. I keep getting an HTTP 400 error saying "Bad Body".  The reason shown is: XAmzContentSHA256Mismatch.

I suspect that this is the issue:
`AWS.Internal.Body.toString >> sha256`

I think that the hash function may have issues with bytes.

**Comment by billstclair** [comment]
I haven't had any coding energy lately. My daughter comes to visit tomorrow, so it may be another week before I manage to look at this. Enjoy the holidays.

**Comment by nicprov** [comment]
No worries, relax and enjoy the holidays! 

I got it mostly working for my needs anyways, but I do think that this would be a feature that people can get some use out of. In all honesty, I can simply make a pull request with the changes that I made and it would be "good enough". The only issue is for the changes on the _elm-aws-core_ repo since the project seems to be dead...

**Comment by billstclair** [comment]
I started looking at this. I'm wondering why you made a `Bytes` version of everything, and copied the code from the non-`Bytes` version. I'm going to try it myself without that, and will likely discover why. But I really HATE copying code, so I'll do my best to avoid it.

And, of course, I need to figure out why the sha256 is failing. The workaround may work well for your encrypted bodies, which already won't decrypt if modified, but it won't work in general.

I sent a PM to rupertlssmith on Elm Slack, to try to bootstrap maintenance for the package.

**Comment by nicprov** [comment]
I agree, code copying is really not ideal. That being said, there are layers of abstraction before an actual HTTP call is made, and each of those components were hard-coded to only work with strings. This makes a rewrite less straightforward which is why I did it the dirty way to get it working for my stuff as quickly as possible.

Regarding the sha256 integrity check, I tried several different things to avoid bypassing the verification, but ultimately failed to get it working. I'll try looking into this more today though.

**Comment by billstclair** [comment]
`Crypto.Hash.sha256` (In `ktonen/elm-crypto`) assumes the string is UTF-8 encoded. It's not hard to fix that, by using `Crypto.SHA.Digest` directly, but it already copies the string before computing the hash, and preventing that requires lots of low-level work, so I think I'll just make a copy.

**Comment by billstclair** [comment]
I submitted a [pull request](https://github.com/ktonon/elm-crypto/pull/9/files) to `ktonen/elm-crypto` to add `Crypto.Hash.shaxxx_bytes` functions, which take `Bytes` as input, instead of `String`. It currently copies the entire contents of the `Bytes`, but I can optimize that later, if two copies in RAM is too big for somebody. I have the changes to `the-sett/elm-aws-core` on my disk, but am waiting for the `ktonen/elm-crypto` changes to be published before proceeding.

I talked with Rupert Smith on Elm Slack. He's happy to accept a pull request for `the-sett/elm-aws-core`.

**Comment by nicprov** [comment]
I tested your changes and the _sha256_bytes_ function works great! I'm not too concerned about performance issues on that particular function, because it's fairly insignificant compared to all the other bottlenecks (encryption, conversion, etc.), but of course, other people may have different use cases.

For reference, I tested it and for a 34MB file, I'm already at ~30 seconds just for the encryption/conversions which is pretty horrible. I didn't notice a difference between using your function vs sending it as "UNSIGNED-PAYLOAD".

And regarding elm-aws-core, that's great news!
